effectful/ops/syntax.py

```diff
@@ -597,6 +624,51 @@ def kwargs(self):
         return self._kwargs
 
 
+class _DataclassTermMeta(type(_BaseTerm)):  # type: ignore
+    def __new__(mcls, name, bases, ns):
+        base_dt = None
+        for b in bases:
+            if dataclasses.is_dataclass(b):
+                base_dt = b
+                break
+
+        if base_dt is None:
+            raise TypeError("No dataclass base found")
```

Comment:
Nit, but this loop only ever sees `(Term, cls)`. You could remove it by considering only that case. `ns` can be simplified similarly.